Can't Get Past the Tutorial: Constant Refreshing Bug

Now that we have ~85 HCDE students using Project Sidewalk for their coursework, we are, unsurprisingly, seeing some new bugs. Here's one that has come up multiple times (that I know of) and it seems very significant (almost a deal breaker for our officially planned Seattle launch this week). Because we don't test the tutorial sequence as much (at least I don't), it's obvious more vulnerable to bugs seeping through.

I have very limited details (what you see below is what I know). I can supply the email addresses for these students so we can follow-up directly. 

Perhaps this is related to https://github.com/ProjectSidewalk/SidewalkWebpage/issues/1563? (If the problem is related to #1563, then perhaps we can have students select a new neighborhood to get around the issue?). **Update**: We've determined that this is not related to #1563.

Student 1:
> I’m trying to do the 3 missions for the Project Sidewalk assignment but every time I try to start exploring and labeling it takes me to the tutorial over and over again. I have tried clicking “have done the tutorial already” button and the page refreshes to the same tutorial page and I’m not sure what to do or how to start. 

Student 2:
> Hello, I do not have a problem with doing the validation, but for the labeling missions, I can't seem to get past the tutorial (done it several times, can't get passed it). Do you know if this is an issue that other people have encountered? The website says it is optimized for chrome and I am doing it in chrome. Any help would be appreciated.

Student 3:
> I was doing the Project Sidewalk assignment and I went through the tutorial, but after I finished it the website wouldn't allow me to do the three labeling missions. Instead it would redirect me to the tutorial over and over again, do you know any solutions to this glitch?

Student 4:
> I have another question, it is about Project Sidewalk. I am trying to explore and mark the curb's, but it keeps sending me back to the tutorial and it won't let me "skip the tutorial" and I am signed on with my UW account. I was able to do the 10 verification's but not the missions. Please let me know what you think, thank you!

Student 5:
> I have been trying to do the 3 labeling missions, but it just makes me do the tutorial over and over. I am signed in and have tried it in multiple browsers with all the security settings off & cookies enabled and still can’t do it. However, the validation works and I did 7 or 8 of those. I also tried creating a different account and that didn’t work either, I do not know if others have been experiencing this, but I do not know how to complete the assignment at the moment though I will try once more on a computer lab computer tomorrow evening.

Student 6:
> I'm trying to do the labeling missions for the Project Sidewalk assignment due tonight but it won't let me get past the tutorial. I've done it three times but after I finish, it leads me back to the tutorial page and won't let me do any other missions.
>
> Have other people been experiencing this difficult? What should I do? 

Student 7:
> I am experiencing issues with the project sidewalk website. I completed the validation missions without any problems, but I cannot access the labeling missions. Each time I try, I am prompted to retake the tutorial. I have completed the tutorial 4 times, but each time I am taken back to the pre-tutorial screen, which I am not able to skip.
> 
> Has anyone else had issues with it? I will do my best to complete the assignment anyway. I would appreciate any insight into the problem.

One unfortunate consequence of this (beyond the inconvenience to the students and the teaching staff--which is significant) is that students are being told to do missions at the UW library. Though this may be a workable solution, those students will be less likely to do more than the minimum missions--which was something I was interested in looking at.

I'm not sure if this is related or not, but I just noticed I had some other errors that popped up in my `sbt` console. This happened sometime during the tutorial:
```
[error] application - No timestamp given for a new label
[error] play - Cannot invoke the action, eventually got an error: org.postgresql.util.PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".
[error] application - 

! @7bgnoo564 - Internal server error, for (POST) [/task] ->

play.api.Application$$anon$1: Execution exception[[PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".]]
	at play.api.Application$class.handleError(Application.scala:296) ~[play_2.10-2.3.8.jar:2.3.8]
	at play.api.DefaultApplication.handleError(Application.scala:402) [play_2.10-2.3.8.jar:2.3.8]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320) [play_2.10-2.3.8.jar:2.3.8]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320) [play_2.10-2.3.8.jar:2.3.8]
	at scala.Option.map(Option.scala:145) [scala-library-2.10.5.jar:na]
Caused by: org.postgresql.util.PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2198) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1927) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:561) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:419) ~[postgresql-9.3-1102-jdbc4.jar:na]
[error] play - Cannot invoke the action, eventually got an error: org.postgresql.util.PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".
[error] application - 

! @7bgnood1m - Internal server error, for (POST) [/task] ->

play.api.Application$$anon$1: Execution exception[[PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".]]
	at play.api.Application$class.handleError(Application.scala:296) ~[play_2.10-2.3.8.jar:2.3.8]
	at play.api.DefaultApplication.handleError(Application.scala:402) [play_2.10-2.3.8.jar:2.3.8]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320) [play_2.10-2.3.8.jar:2.3.8]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320) [play_2.10-2.3.8.jar:2.3.8]
	at scala.Option.map(Option.scala:145) [scala-library-2.10.5.jar:na]
Caused by: org.postgresql.util.PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2198) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1927) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:561) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:419) ~[postgresql-9.3-1102-jdbc4.jar:na]
[error] application - No timestamp given for a new label
[error] play - Cannot invoke the action, eventually got an error: org.postgresql.util.PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".
[error] application - 

! @7bgnool4k - Internal server error, for (POST) [/task] ->

play.api.Application$$anon$1: Execution exception[[PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".]]
	at play.api.Application$class.handleError(Application.scala:296) ~[play_2.10-2.3.8.jar:2.3.8]
	at play.api.DefaultApplication.handleError(Application.scala:402) [play_2.10-2.3.8.jar:2.3.8]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320) [play_2.10-2.3.8.jar:2.3.8]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3$$anonfun$applyOrElse$4.apply(PlayDefaultUpstreamHandler.scala:320) [play_2.10-2.3.8.jar:2.3.8]
	at scala.Option.map(Option.scala:145) [scala-library-2.10.5.jar:na]
Caused by: org.postgresql.util.PSQLException: ERROR: insert or update on table "label" violates foreign key constraint "label_audit_task_id_fkey"
  Detail: Key (audit_task_id)=(15432) is not present in table "audit_task".
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2198) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1927) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:561) ~[postgresql-9.3-1102-jdbc4.jar:na]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:419) ~[postgresql-9.3-1102-jdbc4.jar:na]
```

--- a/public/javascripts/SVLabel/src/SVLabel/data/Form.js
+++ b/public/javascripts/SVLabel/src/SVLabel/data/Form.js
@@ -21,9 +21,7 @@ function Form (labelContainer, missionModel, missionContainer, navigationModel, 
     };
 
     missionModel.on("MissionProgress:complete", function (parameters) {
-        var task = taskContainer.getCurrentTask();
-        var data = self.compileSubmissionData(task);
-        self.submit(data, task);
+        self.submitData(true);
     });
 
     /**
@@ -221,14 +219,13 @@ function Form (labelContainer, missionModel, missionContainer, navigationModel, 
     };
 
     /**
-     * Submit the data
+     * Submit the data via an AJAX post request.
      * @param data
      * @param task
      * @param async
      */
     this.submit = function (data, task, async) {
         if (typeof async === "undefined") { async = true; }
-
         if (data.constructor !== Array) { data = [data]; }
 
         if ('interactions' in data[0] && data[0].constructor === Array) {
@@ -263,8 +260,18 @@ function Form (labelContainer, missionModel, missionContainer, navigationModel, 
 
     $(window).on('beforeunload', function () {
         tracker.push("Unload");
+        self.submitData(false);
+    });
+
+    /**
+     * Manually triggers form submission from other functions.
+     * @param async     Whether data should be submitted asynchronously or not (if undefined,
+     *                  then submits asynchronously by default)
+     */
+    this.submitData = function (async) {
+        if (typeof async === "undefined") { async = true; }
         var task = taskContainer.getCurrentTask();
         var data = self.compileSubmissionData(task);
-        self.submit(data, task, false);
-    });
+        self.submit(data, task, async);
+    }
 }

--- a/public/javascripts/SVLabel/src/SVLabel/data/Form.js
+++ b/public/javascripts/SVLabel/src/SVLabel/data/Form.js
@@ -260,7 +260,22 @@ function Form (labelContainer, missionModel, missionContainer, navigationModel, 
 
     $(window).on('beforeunload', function () {
         tracker.push("Unload");
-        self.submitData(false);
+        var task = taskContainer.getCurrentTask();
+        var data = self.compileSubmissionData(task);
+        var jsonData = JSON.stringify(data);
+
+        // April 17, 2019
+        // What we want here is type: 'appilcation/json'. Can't do that quite yet because the
+        // feature has been disabled, but we should switch back when we can.
+
+        // Source for fix and ongoing discussion is here:
+        // https://bugs.chromium.org/p/chromium/issues/detail?id=490015
+        var headers = {
+            type: 'application/x-www-form-urlencoded'
+        };
+
+        var blob = new Blob([jsonData], headers);
+        navigator.sendBeacon(properties.dataStoreUrl, blob);
     });
 
     /**

--- a/public/javascripts/SVLabel/src/SVLabel/onboarding/Onboarding.js
+++ b/public/javascripts/SVLabel/src/SVLabel/onboarding/Onboarding.js
@@ -499,12 +499,7 @@ function Onboarding(svl, audioEffect, compass, form, handAnimation, mapService, 
             missionContainer.getCurrentMission().setProperty("skipped", true);
         }
         tracker.push('Onboarding_End');
-        console.log("[Onboarding.js] Before setting property");
-        console.log(missionContainer.getCurrentMission().getProperty("isComplete"));
-
         missionContainer.getCurrentMission().setProperty("isComplete", true);
-        console.log("[Onboarding.js] After setting complete property");
-        console.log(missionContainer.getCurrentMission().getProperty("isComplete"));
 
         // Redirects to the audit page and submits all data through Form.js.
         svl.form.submitData(false);

--- a/public/javascripts/SVLabel/src/SVLabel/onboarding/Onboarding.js
+++ b/public/javascripts/SVLabel/src/SVLabel/onboarding/Onboarding.js
@@ -499,9 +499,15 @@ function Onboarding(svl, audioEffect, compass, form, handAnimation, mapService, 
             missionContainer.getCurrentMission().setProperty("skipped", true);
         }
         tracker.push('Onboarding_End');
-        missionContainer.getCurrentMission().setProperty("isComplete", true);
+        console.log("[Onboarding.js] Before setting property");
+        console.log(missionContainer.getCurrentMission().getProperty("isComplete"));
 
-        // Reload the page. This also submits all data through Form.js.
+        missionContainer.getCurrentMission().setProperty("isComplete", true);
+        console.log("[Onboarding.js] After setting complete property");
+        console.log(missionContainer.getCurrentMission().getProperty("isComplete"));
+
+        // Redirects to the audit page and submits all data through Form.js.
+        svl.form.submitData(false);
         window.location.replace('/audit');
     }
 

--- a/public/javascripts/SVValidate/src/data/Form.js
+++ b/public/javascripts/SVValidate/src/data/Form.js
@@ -90,7 +90,20 @@ function Form(url) {
     $(window).on('beforeunload', function () {
         svv.tracker.push("Unload");
         var data = compileSubmissionData();
-        self.submit(data, false);
+        var jsonData = JSON.stringify(data);
+
+        // April 17, 2019
+        // What we want here is type: 'appilcation/json'. Can't do that quite yet because the
+        // feature has been disabled, but we should switch back when we can.
+
+        // Source for fix and ongoing discussion is here:
+        // https://bugs.chromium.org/p/chromium/issues/detail?id=490015
+        var headers = {
+            type: 'application/x-www-form-urlencoded'
+        };
+
+        var blob = new Blob([jsonData], headers);
+        navigator.sendBeacon(properties.dataStoreUrl, blob);
     });
 
     self.compileSubmissionData = compileSubmissionData;

